apex/contrib/csrc/optimizers/multi_tensor_distopt_adam_kernel.cu

```diff
@@ -382,6 +551,51 @@ void multi_tensor_fused_adam_cuda(
   C10_CUDA_CHECK(cudaGetLastError());
 }
 
+void multi_tensor_fused_adam_capturable_cuda(
+  int chunk_size,
+  at::Tensor noop_flag,
+  std::vector<std::vector<at::Tensor>> tensor_lists,  // p_in, m, v, g, p_out
+  at::Tensor grad_scale,
+  at::Tensor lr,
+  float beta1,
+  float beta2,
+  float eps,
+  at::Tensor step,
+  int mode,
+  int bias_correction,
+  float weight_decay)
+{
+  using namespace at;
+
+  // Expect p_in, m, v, g, p_out
+  size_t tl_sz = tensor_lists.size();
+  TORCH_CHECK(tl_sz == 5, "expected tensor lists of size 5");
+  const auto p_in_type = tensor_lists[0][0].scalar_type();
+  const auto g_type = tensor_lists[3][0].scalar_type();
+  const auto p_out_type = tensor_lists[4][0].scalar_type();
+
+  DISPATCH_FLOAT_HALF_AND_BFLOAT(p_in_type, 0, "dist_adam_capturable_cuda_kernel",
+    DISPATCH_FLOAT_HALF_AND_BFLOAT(g_type, 1, "dist_adam_capturable_cuda_kernel",
+      DISPATCH_FLOAT_HALF_AND_BFLOAT(p_out_type, 2, "dist_adam_capturable_cuda_kernel",
```

Comment:
I thought g_type would basically be either `p_in_type` or `p_out_type`, and wondering if it's possible to remove one of these three dispatch's

also, which, p_in or p_out is `model.parameters()`? I'm confused

Reply:
`p_in` corresponds to the master weights (FP32) and `p_out` corresponds to the model weights (BF16).

There are some cases where we don't want to output to `p_out` and I hackily implemented by setting `p_out = p_in`, so it's not guaranteed that `g_type` matches `p_in_type` or `p_out_type`. This is mainly for the FP8 case, and I wanted to keep FP8 logic out of Apex.